src/rust/engine/process_execution/src/sandboxer.rs

```diff
@@ -0,0 +1,290 @@
+            Status::failed_precondition(err_msg)
+        })?;
+        fs::rename(&tmp_dst, &r.dst).map_err(|e| {
+            let err_msg = e.to_string();
+            warn!("Error remaning {} to {}: {}", &tmp_dst, &r.dst, err_msg);
+            Status::failed_precondition(err_msg)
+        })?;
+
+        let confirmation = format!("Copied {} bytes from {} to {}", nbytes, &r.src, &r.dst);
+        info!("{}", confirmation);
+        Ok(Response::new(CopyLocalFileResponse {
+            confirmation: { confirmation },
+        }))
+    }
+}
+
+// The client that sends requests to the sandboxer service.
+pub struct SandboxerClient {
+    grpc_client: SandboxerGrpcClient<Channel>,
+}
+
+impl SandboxerClient {
+    pub async fn connect(socket_path: &Path) -> Result<Self, String> {
+        let socket_path_buf = socket_path.to_path_buf();
+        let channel = Endpoint::try_from("http://[::]:50051")
+            .map_err(|e| e.to_string())?
+            .connect_with_connector(service_fn(move |_: Uri| {
+                let socket_path_buf = socket_path_buf.clone();
+                async {
+                    Ok::<_, std::io::Error>(TokioIo::new(
+                        UnixStream::connect(socket_path_buf).await?,
+                    ))
+                }
+            }))
+            .await
+            .map_err(|e| e.to_string())?;
+        Ok(SandboxerClient {
+            grpc_client: SandboxerGrpcClient::new(channel),
+        })
+    }
+
+    pub async fn copy_local_file(&mut self, src: &Path, dst: &Path) -> Result<String, String> {
+        let src_str = src
+            .to_str()
+            .ok_or(format!("Invalid UTF8 in src: {}", src.to_string_lossy()))?
+            .to_string();
+        let dst_str = dst
+            .to_str()
+            .ok_or(format!("Invalid UTF8 in dst: {}", dst.to_string_lossy()))?
+            .to_string();
+        let request = tonic::Request::new(CopyLocalFileRequest {
+            src: src_str,
+            dst: dst_str,
+        });
+        self.grpc_client
+            .copy_local_file(request)
+            .await
+            .map(|res| res.into_inner().confirmation)
+            .map_err(|e| e.to_string())
+    }
+}
+
+// The spawned sandboxer process and a connection to it.
+pub struct SandboxerProcess {
+    process: ManagedChild,
+    client: SandboxerClient,
+}
+
+// Manages the spawned sandboxer process and communicates with it.
+pub struct Sandboxer {
+    sandboxer_exe: PathBuf,
+    socket_path: PathBuf,
+    process: Option<SandboxerProcess>,
+}
+
+impl Sandboxer {
+    pub async fn new(sandboxer_exe: PathBuf, socket_path: PathBuf) -> Result<Self, String> {
+        Ok(Self {
+            sandboxer_exe,
+            socket_path,
+            process: None,
+        })
+    }
+
+    pub async fn spawn(&mut self) -> Result<(), String> {
+        let mut cmd = Command::new(&self.sandboxer_exe);
+        let cmd = cmd.arg(&self.socket_path);
+        self.process = Some(SandboxerProcess {
```

Comment:
Longer term: Would it be useful for the Pants session or pantsd data structures to keep track of processes on behalf of logic like this? (I'm thinking of test fixtures and other sidecars we may want to spawn in the future. No need to solve now but there is a non-functional capability here in not duplicating process tracking and tear-down logic.)

Reply:
Interesting thought re sidecars. Yes, possibly.